### PR TITLE
[WIP] feat: add CoFiX as a liquidity source

### PR DESCRIPTION
**Pull request by alexkroeger**
## Description

Add CoFiX to asset-swapper.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        view
+        returns (OraclePrice oraclePrice)
+    {
+        (uint32 k, uint32 updatedAt, uint32 theta) = ICoFiXController(_getCoFiXControllerAddress()).getKInfo(tokenAddress);
+        (uint256 ethAmount, uint256 erc20Amount, uint256 blockNum); = INestOracle(_getNestOracleAddress()).checkPriceNow(tokenAddress);
+
+        uint k2 = uint(k);
+        uint theta2 = uint(theta);
+
+        return [ethAmount, erc20Amount, blockNum, k2, theta];
```
Comment (alexkroeger) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:209]: theta -> theta2

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                (didSucceed, bytes memory resultData) =
+                    address(makerTokenPair).staticcall.gas(COFIX_CALL_GAS)(
+                        abi.encodeWithSelector(
+                            ICoFiXPair(0).calcOutToken1.selector,
+                            takerTokenAmounts[i],
+                            _getOraclePrice(makerToken)
+                    ));
+                makerTokenAmounts[i] = abi.decode(resultData, (uint256[]))[0];
+            } else {
+                // if it's converting a non-WETH token to another non-WETH token
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:90]: I think it's unlikely we'd want this case. From what I can see everything is ETH/X.

MultiHop should take care of this when it is X/Y rather then us having to do it here and in the bridge.

We do this for Kyber as all of Kyber trades are actually X->ETH->Y.

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    )
+        public
+        view
+        returns (uint256[] memory makerTokenAmounts)
+    {
+        // make sure it's not WETH-WETH
+        _assertValidPair(makerToken, takerToken);
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
+        ICoFiXPair takerTokenPair = takerToken == _getWethAddress() ?
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:63]: This appears to be similar to DODO where there is a "base" token (token0) and a quote token (token1). And we need to get the oracle data of the "quote" token (token1). And branch off the selector based on the quote/base.

Thhen maybe we could just do:

```
for (uint256 i = 0; i < numSamples; i++) {
...
                (didSucceed, bytes memory resultData) =
                    address(pool).staticcall.gas(COFIX_CALL_GAS)(
                        abi.encodeWithSelector(
                            takerToken == quoteToken ? ICoFiXPair(0).calcOutToken0.selector : ICoFiXPair(0).calcOutToken1.selector,
                            takerTokenAmounts[i],
                            _getOraclePrice(quoteToken)
                    ));
                makerTokenAmounts[i] = abi.decode(resultData, (uint256[]))[0];
```

I may have got the 0 and 1 around the wrong way though.

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    /// @dev Returns the oracle information from NEST given a token
+    /// @param tokenAddress Address of the token contract.
+    /// @return OraclePrice NEST oracle data
+    function _getOraclePrice(address tokenAddress)
+        private
+        view
+        returns (OraclePrice oraclePrice)
+    {
+        (uint32 k, uint32 updatedAt, uint32 theta) = ICoFiXController(_getCoFiXControllerAddress()).getKInfo(tokenAddress);
+        (uint256 ethAmount, uint256 erc20Amount, uint256 blockNum); = INestOracle(_getNestOracleAddress()).checkPriceNow(tokenAddress);
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:204]:
```suggestion
        (oraclePrice.ethAmount, oraclePrice.erc20Amount, oraclePrice.blockNum); = INestOracle(_getNestOracleAddress()).checkPriceNow(tokenAddress);
```

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+    /// @dev Returns the oracle information from NEST given a token
+    /// @param tokenAddress Address of the token contract.
+    /// @return OraclePrice NEST oracle data
+    function _getOraclePrice(address tokenAddress)
+        private
+        view
+        returns (OraclePrice oraclePrice)
+    {
+        (uint32 k, uint32 updatedAt, uint32 theta) = ICoFiXController(_getCoFiXControllerAddress()).getKInfo(tokenAddress);
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:203]: maybe something like 
```suggestion
        (oraclePrice.K, uint32, oraclePrice.theta) = ICoFiXController(_getCoFiXControllerAddress()).getKInfo(tokenAddress);
```

File: packages/asset-swapper/contracts/src/interfaces/ICoFiXFactory.sol [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+
+
+interface ICoFiXFactory {
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/interfaces/ICoFiXFactory.sol:22]: nit: I just inline these above the `CoFiXSampler`, nothing else is going to use it.

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    /// @return makerTokenAmounts Maker amounts bought at each taker token
+    ///         amount.
+    function sampleSellsFromCoFiX(
+        address takerToken,
+        address makerToken,
+        uint256[] memory takerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory makerTokenAmounts)
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:57]: Probably want to return the pool here so it can be encoded in the FillData.

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts [hunk 8]
```diff
@@ -907,6 +907,32 @@ export class SamplerOperations {
...
+    public getCoFiXBuyQuotes(
+        makerToken: string,
+        takerToken: string,
+        makerFillAmounts: BigNumber[],
+    ): SourceQuoteOperation {
+        return new SamplerContractOperation({
+            source: ERC20BridgeSource.CoFiX,
+            contract: this._samplerContract,
+            function: this._samplerContract.sampleBuysFromCoFiX,
+            params: [takerToken, makerToken, makerFillAmounts],
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:932]: Probably want to manually parse this (see `callback`) and extract the `pool` in the FillData.

**Review by dekz — COMMENTED** [review]
Thiink this needs to be rebased as well, we're on solidity 0.6 now.

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinCoFiX.sol [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        bytes memory bridgeData
+    )
+        internal
+        returns (uint256 boughtAmount)
+    {
+        (address fromTokenAddress, uint256 fee) = abi.decode(bridgeData, (address, uint256));
+
+        // Grant the CoFiX Router contract an allowance to sell the first token.
+        if (fromTokenAddress == address(WETH)) {
+            // TODO we may be able to get away with Pair direct trade and avoid WETH/deposit withdraw
```
Comment (dekz) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinCoFiX.sol:81]: check if we have access to go direct and skip the weth deposit/withdraw

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        bool didSucceed;
+        bytes memory resultData;
+        fee = INestOracle(_getNestOracleAddress()).checkPriceCost();
+        for (uint256 i = 0; i < numSamples; i++) {
+            if (makerToken == _getWethAddress()) {
+                // if converting to WETH, sell all of the token in the taker token pair
+                (didSucceed, resultData) =
+                    address(takerTokenPair).staticcall.gas(COFIX_CALL_GAS)(
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:81]: change to try/catch

File: contracts/utils/contracts/src/DeploymentConstants.sol [hunk 11]
```diff
@@ -60,6 +60,14 @@ contract DeploymentConstants {
...
     /// @dev Mainnet address of the DODO Helper contract
     address constant private DODO_HELPER = 0x533dA777aeDCE766CEAe696bf90f8541A4bA80Eb;
+    /// @dev Mainnet address of the CoFiX Router contract
+    address constant private COFIX_ROUTER_ADDRESS = 0x26aaD4D82f6c9FA6E34D8c1067429C986A055872;
+    /// @dev Mainnet address of the CoFiX factory contract
+    address constant private COFIX_FACTORY_ADDRESS = 0x66C64ecC3A6014733325a8f2EBEE46B4CA3ED550;
+    /// @dev Mainnet address of the CoFiX Controller contract
+    address constant private COFIX_CONTROLLER_ADDRESS = 0xd694eF4C82E50F3f184877572dCD6A45B8FEae87;
+    /// @dev Mainnet address of the NEST Oracle contract
+    address constant private NEST_ORACLE_ADDRESS = 0x94F36FAa6bB4f74009637292b09C355CcD3e80Eb;
```
Comment (dekz) [inline contracts/utils/contracts/src/DeploymentConstants.sol:70]: revert these changes, preferring AS

File: packages/asset-swapper/contracts/src/CoFiXSampler.sol [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        bool didSucceed;
+        bytes memory resultData;
+        fee = INestOracle(_getNestOracleAddress()).checkPriceCost();
+        for (uint256 i = 0; i < numSamples; i++) {
+            if (makerToken == _getWethAddress()) {
+                // if converting to WETH, find out how much of the taker token to send in
+                (didSucceed, resultData) =
+                    address(takerTokenPair).staticcall.gas(COFIX_CALL_GAS)(
```
Comment (dekz) [inline packages/asset-swapper/contracts/src/CoFiXSampler.sol:143]: TODO don't seem to get buys, probably got the order wrong

File: packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 13]
```diff
@@ -100,6 +100,16 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
         if (isFromETH) {
             ethAmount = ethAmount.plus(sellAmount);
         }
+
+        // TODO really hate this, either tally it into `protocolFee`? or
+        // have the protocol fee in the Native fillData? and tally all `feeInWei`
+        // Add additional ETH fees
+        quote.orders.forEach(o => {
```
Comment (dekz) [inline packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts:107]: Find a cleaner solution to this

File: packages/asset-swapper/src/utils/swap_quote_calculator.ts [hunk 14]
```diff
@@ -134,12 +135,19 @@ export class SwapQuoteCalculator {
...
             ...opts,
-            feeSchedule: _.mapValues(opts.feeSchedule, gasCost => (fillData?: FillData) =>
-                gasCost === undefined ? 0 : gasPrice.times(gasCost(fillData)),
-            ),
+            feeSchedule: _.mapValues(opts.feeSchedule, (gasCost, source) => (fillData?: FillData) => {
+                if (gasCost === undefined) {
+                    return 0;
+                }
+                let gasFee = gasPrice.times(gasCost(fillData));
+                if (source === ERC20BridgeSource.CoFiX) {
```
Comment (dekz) [inline packages/asset-swapper/src/utils/swap_quote_calculator.ts:146]: TODO find a cleaner solution to this
